src/Nethermind/Nethermind.Network.Discovery/Serializers/DiscoveryMsgSerializerBase.cs

```diff
@@ -45,62 +47,79 @@ protected byte[] Serialize(byte type, Span<byte> data)
         Span<byte> resultSpan = result.AsSpan();
         resultSpan[32 + 65] = type;
         data.CopyTo(resultSpan.Slice(32 + 65 + 1, data.Length));
-        
+
         Span<byte> payload = resultSpan.Slice(32 + 65);
         Keccak toSign = Keccak.Compute(payload);
         Signature signature = _ecdsa.Sign(_privateKey, toSign);
         signature.Bytes.AsSpan().CopyTo(resultSpan.Slice(32, 64));
         resultSpan[32 + 64] = signature.RecoveryId;
-            
+
         Span<byte> forMdc = resultSpan.Slice(32);
         ValueKeccak mdc = ValueKeccak.Compute(forMdc);
         mdc.BytesAsSpan.CopyTo(resultSpan.Slice(0,32));
         return result;
     }
 
-    protected (PublicKey FarPublicKey, byte[] Mdc, byte[] Data) PrepareForDeserialization(byte[] msg)
+    protected (PublicKey FarPublicKey, byte[] Mdc, IByteBuffer Data) PrepareForDeserialization(IByteBuffer msg)
     {
-        if (msg.Length < 98)
+        if (msg.ReadableBytes < 98)
         {
             throw new NetworkingException("Incorrect message", NetworkExceptionType.Validation);
         }
-        
-        byte[] mdc = msg.Slice(0, 32);
-        Span<byte> signature = msg.AsSpan(32, 65);
-        // var type = new[] { msg[97] };
-        byte[] data = msg.Slice(98, msg.Length - 98);
-        Span<byte> computedMdc = ValueKeccak.Compute(msg.AsSpan(32)).BytesAsSpan;
-        
+        IByteBuffer data = msg.Slice(98, msg.ReadableBytes - 98);
+        Span<byte> msgBytes = msg.ReadAllBytes().AsSpan();
+        Span<byte> mdc = msgBytes.Slice(0, 32);
+        Span<byte> sigAndData = msgBytes.Slice(32);
+        Span<byte> computedMdc = ValueKeccak.Compute(sigAndData).BytesAsSpan;
+
         if (!Bytes.AreEqual(mdc, computedMdc))
         {
             throw new NetworkingException("Invalid MDC", NetworkExceptionType.Validation);
         }
-        
-        PublicKey nodeId = _nodeIdResolver.GetNodeId(signature.Slice(0, 64).ToArray(), signature[64], msg.AsSpan(97, msg.Length - 97));
-        return (nodeId, mdc, data);
+
+        PublicKey nodeId = _nodeIdResolver.GetNodeId(sigAndData.Slice(0, 64).ToArray(), sigAndData[64], sigAndData.Slice(65, sigAndData.Length - 65));
+        return (nodeId, mdc.ToArray(), data);
```

Comment:
Do we have unit tests covering this logic?

Reply:
Directly testing this logic -> no
but this function is used in deserializing a lot of messages, so if we are able to deserialize those messages using this logic - that proves that this logic is also correct?